library/alloc/src/vec.rs

```diff
@@ -2391,11 +2384,62 @@ where
+    fn spec_extend(&mut self, iterator: IntoIter<T>) {
+        // Avoid reallocation if we can use iterator's storage instead. This requires 1 memcpy and 0-1 memmove
+        // while reallocation would require 1 alloc, 1-2 memcpy, 1-2 free.
+        //
+        // # illustration of some extend scenarios (not exhaustive)
+        //
+        //   == step ==              == memory ==              == self ==         == iter / v ==
+        //                   0123456789abcdef0123456789abcdef
+        //                   0---------------1---------------
+        //
+        // ## non-empty self, partially consumed iterator
+        //
+        //    [initial]      AAAA_-----__BBB___--------------  Vec(0x00, 4, 5)    IntoIter(0x0a, 0x0c, 0x0f, 8)
+        // ³  into_vec       AAAA_-----____BBB_--------------  Vec(0x00, 4, 5)    Vec(0x0a, 7, 8)
+        // ²  prepend        _____-----AAAABBB_--------------  Vec(0x00, 0, 5)    Vec(0x0a, 7, 8)
+        // ⁴  *self = v      ----------AAAABBB_--------------  Vec(0x0a, 7, 8)
+        //
+        // ## empty self, partially consumed iterator
+        //
+        //    [initial]      ____------__BBBB__--------------  Vec(0x00, 0, 4)    IntoIter(0x0a, 0x0c, 0x10, 8)
+        // ³  into_vec       ____------BBBB____--------------  Vec(0x00, 0, 4)    Vec(0x0a, 4, 8)
+        // ⁴  *self = v      ----------BBBB____--------------  Vec(0x0a, 4, 8)
+        //
+        // ## empty self, pristine iterator
+        //
+        //    [initial]      ----------BBBB____--------------  Vec(0x00, 0, 0)    IntoIter(0x0a, 0x0a, 0x0e, 8)
+        //    *self = v      ----------BBBB____--------------  Vec(0x0a, 4, 8)
+        //
+        // ## insufficient capacity
+        //
+        //    [initial]      AAAAA-----BBBBBB__--------------  Vec(0x00, 5,  5)   IntoIter(0x0a, 0x0a, 0x0f, 8)
+        // ¹² reserve(6)     ----------BBBBBB__--AAAAA______-  Vec(0x14, 5,  11)  IntoIter(0x0a, 0x0a, 0x0f, 8)
+        // ²  ptr:copy_n     ----------________--AAAAABBBBBB-  Vec(0x14, 11, 11)  IntoIter(0x0a, 0x0f, 0x0f, 8)
+        // ⁴  drop           --------------------AAAAABBBBBB-  Vec(0x14, 11, 11)
+        //
+        //  ¹ malloc
+        //  ² memcpy
+        //  ³ memmove
+        //  ⁴ free
+        //
+        if mem::size_of::<T>() > 0
+            && self.capacity() - self.len() < iterator.len()
+            && iterator.cap - iterator.len() >= self.len()
+        {
+            // Safety: we just checked that IntoIter has sufficient capacity to prepend our elements.
+            // Prepending will then fill the uninitialized prefix.
+            let v = unsafe {
+                let mut v = iterator.into_vec_with_uninit_prefix(self.len() as isize);
+                ptr::copy_nonoverlapping(self.as_ptr(), v.as_mut_ptr(), self.len);
+                self.set_len(0);
+                v
+            };
+            *self = v;
+            return;
         }
-        iterator.ptr = iterator.end;
+        iterator.move_into(self);
```

Comment:
It would be cooler if the code are linked to the cool diagram.
```suggestion
        // Insufficient capacity
        iterator.move_into(self);
```

Reply:
Well, yes and no. 

With the optimization present that is indeed all it covers. But it is the general codepath that also works without the optimization. So I don't want to give the impression that it can only handle that case.